### ISSUE TITLE
Quick right-click marquee/lasso selection (and revert back to previous tool upon deselect)

### ISSUE THREAD

**Issue by awesomez** [aseprite/aseprite#1792]
I'm positive this has been requested before, but I'm so frustrated, I'm going to request it again. 

The problem is, I'm struggling heavily with animation in Aseprite against my speedy animation workflows in Graphics Gale -- I really *really* want to like Aseprite for animation, but I can't right-click and drag a marquee selection to move parts about the image, then quickly touch them up with the brush tool without moving my hand all around the keyboard to swap tools.

Graphics Gale solved this issue by letting the user right-click anywhere in the canvas and either select a color, or drag a marquee selection with the right-mouse button (then alter that object's position/rotation immediately). Once you were done (i.e. clicked off the selection), it automatically switched you back to your previous tool.

Without my mouse-based workflow, making pixel-animation in anything else but Graphics Gale is sooo painful it makes me want to cry. 

Unfortunately, since I'm an animator, I can't yet bring myself to switch entirely to Aseprite because of this (huge) workflow hitch, and I'm afraid I'll end up regressing (yet again) back to Graphics Gale for everything pixel-related simply because it does almost everything I need very well -- but without the bells and whistles of Aseprite of course (such as Rotsprite and Pixel-perfect drawing.) D:

Can you *please* fix this workflow issue? -- I *love* my mouse-based workflow in Graphics Gale, but I also love Aseprite for everything else it does -- _except_ for when it makes me cross my arms to find all the keyboard shortcuts over and over for common animation operations.

The beauty of solving this issue for animation is that it will benefit sprite modification for static images too. I am currently in the process of shifting around parts of sprites to tweak size/proportions of similar-looking characters and this is immensely hard in Aseprite with all the back-and-forth keyboard shortcuts.
This marquee/lasso-tool tweak would make this process SO much easier.

**Comment by KashouC** [comment]
I'm not arguing for or against this change, it doesn't matter to me and I understand the frustration of having to move all over the keyboard, I just wanted to suggest taking some time to remap the tools you use to the left side of the keyboard. The only time I ever have to move my hand is for Shift+U which is the hue/saturation adjustment tool.

I even bound things like invert to 3, delete to 4 and increase/decrease brush to 1 and 2.

That said I'm sure a mode for this would be helpful too.

**Comment by awesomez** [comment]
Fair enough -- I never considered moving the shortcuts to one side of the keyboard, but it makes sense. As I said, I got frustrated trying to animate with the default (keyboard-based) setup and wasn't thinking. I suppose I could do what you suggested in the meantime to keep the keyboard painfulness at a minimum while this gets implemented in some form. 

The major thing to me is that it requires an extra two actions from me (a set of actions which are way slower, cumulatively, than my current workflow.) 

The sole "reason" of having to press the key is to swap to the marquee, but I also have to click/drag to select, and _then_ move/rotate/scale the part (the _only_ thing I'm trying to do), and then manually swap back to my brush tool after making some tweaks, which will likely require me to swap back over to the marquee again and rinse/repeat a few more times if the changes don't look quite right in the animation. This is especially cumbersome when working with larger sprites that need lots of adjustment of these parts. Graphics Gale makes working with large sprites like this just as easy as small ones because of the mouse-based action is so accessible to the user.

Perhaps a separate brush/marquee (and brush/lasso) hybrid tool would be easier? 

This would let one pick colors and perform the marquee/lasso function with the right-click and draw with left-click. The difference would likely be that adding/removing additional photoshop-style selections (in this mode) wouldn't be possible. And I'd be okay with that. -- If I wanted that kind of flexibility with my marquee, I'd go grab the marquee tool itself.

**Comment by dacap** [comment]
Related to #967 and #1197.

Let me see if I get it right, e.g. when we have the Pencil tool selected:

1. If you left-click/drag with left mouse button, you paint with the Foreground color.
2. If you right-click one pixel (mouse button down and up in the same pixel), it uses Eyedropper for that pixel and you are still in Pencil tool to draw with left mouse button.
3. If you push right mouse button, drag it, and release it, it should behave like the Rectangular Marquee tool, temporarily switch to the Rectangular Marquee, until you right-click again one pixel (which should only deselect and go back to the Pencil tool, without calling the Eyedropper)
4. Something similar could be done with Ctrl+right mouse button and the Lasso tool as specified in #1197.

**Comment by awesomez** [comment]
One slight correction to #3 really quick:

3. If you push right mouse button, drag it, and release it, it should behave like the Rectangular Marquee tool, temporarily switch to the Rectangular Marquee, **until you right-click again one pixel** (which should only deselect and go back to the Pencil tool, without calling the Eyedropper)

That should be:

3. If you push right mouse button, drag it, and release it, it should behave like the Rectangular Marquee tool, temporarily switch to the Rectangular Marquee, **until you left-click again one pixel outside of the selection or right-click inside the selection** (which should only deselect and go back to the Pencil tool, without calling the Eyedropper)

But yep -- that's exactly what I'd be describing there.

**Comment by dacap** [comment]
So when you use right mouse button to do a rectangular selection, after releasing the right mouse button you start dragging the selection, it looks like there is no chance for scaling/rotation (as you said in a previous post). Can you confirm if it is only translation? I guess that the selection is dragged until you press the left mouse button or the right mouse button again. (or I don't know if they have different behaviors).

I really have to install GraphicsGale to check how this mode works.

Anyway there are a lot of possible combinations for this kind of stuff, this might be a possible candidate for #1403

**Comment by dacap** [comment]
A summary of this feature:

Without selection:
1. left click paints with fg color
1. right click pick colors
1. right drag behaves as the rectangular marquee

With selection:
1. left drag moves/transform the selection
2. right click deselects and goes back to the previous tool (keep track of the last non-selection tool)
3. right drag deselects and behaves as the rectangular marquee

This is a special mode (not only right-click), that would completely disable the possibility to paint inside the selection only.

**Comment by awesomez** [comment]
This definitely sounds workable, @dacap. 

The idea, for me at least, is to be able to quickly move around parts/chunks of a sprite (for the purposes of animation), then immediately tweak or redraw some parts of them (after they were shifted and rotated). 

This speeds up the animation process greatly when only doing minor tweaks to already-created portions of a sprite (i.e. for a face or head or for higher-resolution big muscular character throwing a punch or moving tiny accessories or large guns around slightly).
